### PR TITLE
Move crate drop-down to search results page

### PR DESCRIPTION
**Pull request by jsha**
This reduces clutter on doc pages.

Part of #59840

r? @GuillaumeGomez 

Demo: https://rustdoc.crud.net/jsha/crates-in-results/std/index.html?search=str

### CODE REVIEW

File: src/librustdoc/html/static/js/search.js [hunk 1]
```diff
@@ -1126,9 +1126,16 @@ window.initSearch = function(rawSearchIndex) {
             }
         }
 
-        var output = "<h1>Results for " + escape(query.query) +
+        let crates = `<select id="crate-search"><option value="All crates">All crates</option>`;
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/js/search.js:1129]: Isn't eslint complaining when you use backticks without `${}` content?

**Comment by GuillaumeGomez** [comment]
As said previously, I really like it! The display needs a bit of improvement: the `in` is too small I think (but maybe it's voluntary?). Maybe we should force the dropdown to be on its own line on mobile? Currently it gives (with long search string):

Screenshot from 2022-01-02 18-09-51

The style of the dropdown is incomplete I think: the border is looking strange.

In any case, it's a nice start. :)

File: src/librustdoc/html/static/js/search.js (hunk 1 above)
Comment (jsha) [reply src/librustdoc/html/static/js/search.js:1129]: Nope! I wouldn't expect it to. Backticks do a bunch of things besides template expansion. They also allow multiline strings, and serve as a different delimiter so you don't have to backslash double quotes (`\"`)

**Comment by jsha** [comment]
> the in is too small I think (but maybe it's voluntary?).

This was on purpose, since `in ____` isn't part of the title (`<h1>`). But I could make it the same size as the `<h1>`, just outside the tag? I don't have a strong opinion about what's better.

> Maybe we should force the dropdown to be on its own line on mobile? 

Good idea.

**Comment by camelid** [comment]
I also had an idea recently that could help with this (I forget if I mentioned it or not). We could add a `crate:some_crate` search syntactical form. Then we could add it as an option to the advanced search builder @jsha proposed.

**Comment by camelid** [comment]
> This was on purpose, since `in ____` isn't part of the title (`<h1>`). But I could make it the same size as the `<h1>`, just outside the tag? I don't have a strong opinion about what's better.

It looks very strange to me. The `select` also renders differently from how it used to:

image

The bezel seems a bit extreme.

**Comment by jsha** [comment]
Interesting! What platform is that?

**Comment by camelid** [comment]
Firefox on macOS.

**Comment by jsha** [comment]
Alright, I've increased the size of `in` and fixed the border issue (I had overzealously removed a `border:0` in the CSS that turned out to be needed). I also tweaked the padding slightly. Demo is pushed and ready for another look!

File: src/librustdoc/html/static/js/search.js [hunk 2]
```diff
@@ -1126,9 +1126,16 @@ window.initSearch = function(rawSearchIndex) {
...
+        for (let c of window.ALL_CRATES) {
+            crates += `<option value="${c}" ${c == filterCrates && "selected"}>${c}</option>`;
+        }
+        crates += `</select>`;
+        var output = `<div id="search-settings">
+            <h1 class="search-results-title">Results for ${escape(query.query)} ` +
             (query.type ? " (type: " + escape(query.type) + ")" : "") + "</h1>" +
-            "<div id=\"titles\">" +
+            ` in ${crates} ` +
+            "</div><div id=\"titles\">" +
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/js/search.js:1138]: I think backticks would be better here as well.

File: src/ci/docker/host-x86_64/mingw-check/Dockerfile [hunk 3]
```diff
@@ -40,5 +40,5 @@ ENV SCRIPT python3 ../x.py --stage 2 test src/tools/expand-yaml-anchors && \
            /scripts/validate-toolstate.sh && \
            /scripts/validate-error-codes.sh && \
            # Runs checks to ensure that there are no ES5 issues in our JS code.
-           es-check es5 ../src/librustdoc/html/static/js/*.js && \
+           es-check es6 ../src/librustdoc/html/static/js/*.js && \
```
Comment (GuillaumeGomez) [inline src/ci/docker/host-x86_64/mingw-check/Dockerfile:43]: We wanted to make this change in bigger PR but I guess it's fine... Don't forget to update the comment just above please.

Comment (jsha) [reply src/ci/docker/host-x86_64/mingw-check/Dockerfile:43]: I actually think a giant PR adding ES6 features wherever possible, and updating the linter at the same time would be a bad idea. The way we should do it is like I'm doing here:

 - update the linter to accept es6 as a standalone commit
 - start using es6 features where they are useful
 - do a series of standalone commits that do one of two things:
   - adopt an ES6 feature where it is useful (e.g. const everything that can be consted), or
   - migrate a discrete chunk of code to using various useful ES6 features.

Comment (GuillaumeGomez) [reply src/ci/docker/host-x86_64/mingw-check/Dockerfile:43]: Like I said, it's fine. :wink:

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Thanks!

@bors: r+

**Pull request merged**